include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +38,88 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(void*));
+    static_assert(alignof(sycl::queue) >= alignof(void*));
+
+    std::byte __buf[sizeof(sycl::queue)];
+
+    bool __has_queue() const
+    {
+        bool res = (nullptr != reinterpret_cast<void* const&>(*this));
+        std::atomic_signal_fence(std::memory_order_acq_rel); // mitigate possible reordering due to type punning
+        return res;
+    }
+
+  public:
+    template <typename... _Args>
+    __queue_holder(_Args... __args)
+    {
+        new (this) sycl::queue(__args...);
+    }
+
+#if _ONEDPL_PREDEFINED_POLICIES
+    __queue_holder(__global_instance_tag)
+    {
+        if (!sycl::device::get_devices().empty())
+            new (this) sycl::queue;
+        else
+        {
+            // an "impossible" case of SYCL providing no devices, which we however must handle
+            std::memset(this, 0, sizeof(void*));
+            // Since the queue holder does not have a valid queue in this case,
+            // the predefined device policies are de-facto unusable.
+            // That seems OK though, because there is no device to run SYCL anyway.
+        }
+    }
+#endif
+
+    ~__queue_holder()
+    {
+        if (__has_queue())
+            __queue_ref().~queue();
+    }
+
+    // Copy and move operations have to be explicit
+    __queue_holder(const __queue_holder& __h) { new (this) sycl::queue(__h.__queue_ref()); }
+    __queue_holder(__queue_holder&& __h)      { new (this) sycl::queue(std::move(__h.__queue_ref())); }
+
+    __queue_holder& operator=(const __queue_holder& __h)
+    {
+        if (this != &__h)
+            __queue_ref() = __h.__queue_ref();
+        return *this;
+    }
+
+    __queue_holder& operator=(__queue_holder&& __h)
+    {
+        if (this != &__h)
+            __queue_ref() = std::move(__h.__queue_ref());
+        return *this;
+    }
+
+    const sycl::queue& __queue_ref() const
+    {
+        assert(__has_queue());
+        return reinterpret_cast<const sycl::queue&>(*this);
+    }
+    sycl::queue& __queue_ref()
+    {
+        assert(__has_queue());
```

Comment:
What this assertions will doing in release mode? Looks like as crash in release ma happen later...

Reply:
May be better to throw some exception instead of `assert` ?

Reply:
One of the design goals for the patch was to make it in a "pay-as-you-go" way - i.e. avoid having any impact from the workaround to the normal usage of valid device policies. I have to be convinced that throwing an exception for the sake of better diagnostic for a strange corner case is worth even that small extra check on the regular code path.

Moreover, if that extra check is deemed worthwhile, we might as well not throw anything but try creating/using a valid queue - again, see https://github.com/oneapi-src/oneDPL/compare/main...b83ab2d for a possible implementation.

Reply:
Sure, I understand.  This is certainly an edge case, and to have all of oneDPL pay some penalty (even if it quite small) would require some real justification.  One could argue that it would not take much time to diagnose an error which is only present when no valid device is available, even if this worsens the message as compared to the original experience.

Reply:
I often run into the situation where I am attempting to run a oneDPL benchmark / test when there is not a valid SYCL device available. This is due to testing multiple drivers / platforms and something not being installed or configured correctly on the system. Seeing the current error of `PI_ERROR_DEVICE_NOT_FOUND` makes it clear what the issue is and I believe that changing this behavior to instead be a segmentation fault could result in a worse experience.

If I look at it from the perspective of someone that is trying oneDPL for the first time and their system happens to be not configured correctly to make a SYCL device available, encountering a segmentation fault instead of a clearer error message might end up scaring that person away.

Reply:
> If I look at it from the perspective of someone that is trying oneDPL for the first time and their system happens to be not configured correctly to make a SYCL device available, encountering a segmentation fault instead of a clearer error message might end up scaring that person away.

I think it can be mitigated by building it in DEBUG mode. Not super convenient but also not the end of the world. Anyway, if `default_selector_v` doesn't return a device we cannot rely on the SYCL implementation throwing an exception because strictly speaking, it's Undefined Behavior from the SYCL spec standpoint. We, of course, can invent our own exception but "should we?" is a big question to me because it increases the size of binary code and it's not zero-cost.

Reply:
I've attempted to improve the assertion diagnostics with a more descriptive predicate name and an attached message.
@SergeyKopienko @adamfidel @danhoeflinger please take a look again.